### ISSUE TITLE
YamlFileManager should not rely on ConfigurationResourceResolver

### ISSUE THREAD

**Issue by fuss86** [strongbox/strongbox#1169]
# Task Description

`org.carlspring.strongbox.yaml.YamlFileManager` classes are too complex and should be simplified.

We should benefit from [spring boot configuration order](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html) and keep the defaults in the application.properties (or YAML equivalent, see #1167).

# Tasks

* [x] Get rid of `org.carlspring.strongbox.yaml.YamlFileManager#getPropertyKey` and `org.carlspring.strongbox.yaml.YamlFileManager#getDefaultLocation`. `org.carlspring.strongbox.yaml.YamlFileManager` and all of its kids should not rely/depend on `org.carlspring.strongbox.resource.ConfigurationResourceResolver`. 
* [x] use `@Value` annotation (or better alternative from `Spring`/`Spring Boot`) to server the `Resource`. 
* [x] `org.carlspring.strongbox.yaml.YamlFileManager#getResource` probably should be `abstract` (or propose some better alternative)

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

**Comment by ajaybhat** [comment]
I would like to work on this.

**Comment by carlspring** [comment]
@ajaybhat ,

Thanks! Feel free to have a look at it and to propose a pull request for it! :)

**Comment by bono007** [comment]
@carlspring I am interested in this one - while I wait on outcome of 1213 I will take this one on.

**Comment by carlspring** [comment]
Go for it! :)

**Comment by bono007** [comment]
Improvement for later: `PropertiesBooter.java` can be replaced w/ a `@ConfigurationProperties` prefixed on `strongbox.` and then the default system property setting portion should be replaced with an ApplicationContextInitializer to be on the same sife. `@PostConstruct` may be too late to set system properties.

**Comment by carlspring** [comment]
@strongbox/core-developers , @ptirador ,

What do you guys think?

**Comment by bono007** [comment]
Some benefits of type safe configuration properties are validation, relaxed binding, better support in ide on clicking back forth between props class and props file. https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties

**Comment by fuss86** [comment]
Yes, I like this idea. 👍

**Comment by ptirador** [comment]
I like it too :+1:

**Comment by carlspring** [comment]
@bono007 ,

Please, feel free to proceed!

**Comment by bono007** [comment]
Shall I keep the work under this ticket or create another and link it?

**Comment by carlspring** [comment]
I think you can use this one.

**Comment by bono007** [comment]
Oh my... after digging in on my own suggestion...



So in regards to moving `PropertiesBooter` over to `@ConfigurationProperties` its not a good fit because...

1. The current properties have multiple deep paths off of `strongbox.` such as `strongbox.nuget.download.feed` and `strongbox.storage.booter.basedir`
1. The object property names do not match the actual property names (for ex: `getHomeDirectory` is mapped to `strongbox.home`)

ConfigurationProperties has a terrible story when it comes to supporting the above. 
* For nested properties it requires an inner class per dot-segment in the path which would result in a deep, nasty, nested object graph in the properties class. 
* It does not support re-mapping names and relies on direct 1-1 naming convention between property name and field name. 

This is what it would look like to map `strongbox.nuget.download.feed`...

```
class StrongboxProperties {
	static class Nuget {
		static class Download {
			private String feed;
			...
		}
	}
}
```

And to maintain the property name `strongbox.home` the accessor would have to be `getHome` instead of the current `getHomeDirectory`.

In summary, `@ConfigurationProperties` is useable for properties if they are designed up front w it in mind but terrible in switching over after the fact. 

I am still looking at moving over the `@PostConstruct` to an ACI.

**Comment by carlspring** [comment]
To be honest, I'm not a massive fan of all the empty classes used for pretty much nothing...

Perhaps, in this case we should park this task...? @strongbox/core-developers , @ptirador what do you guys think.

**Comment by bono007** [comment]
Yeh i think the switch to configprops is not a good one after i dug into it. The latter part of moving the system prop setting into an app context initializer is still valid though. 

I have a PR in for that now.

Once that PR is done i will get back to the initial motivation of this ticket - simplifying that resource loader and yaml props.

**Comment by bono007** [comment]
So for the YamlFileManager and its kids, the behavior for current resource loading is this (using UsersFileManager values as an example - all of the kids follow this format)

1) If the "strongbox.users.config.yaml" property is set use its value as the path to load the yaml resource with

2) Otherwise, if there is an actual file at "$HOME/etc/conf/strongbox-security-users.yaml" load the yaml from there.

3) Otherwise, load the yaml from "etc/conf/strongbox-security-users.yaml".

I am trying to understand the need for 2) and 3). That is what is making it a bit custom. If we just used the property value or a single default location this becomes as simple as the following in the kids...

```
@Value("${strongbox.users.config.yaml:${home}/etc/conf/strongbox-security-users.yaml}")
private Resource resource;
```

The configuration load ordering referenced in the ticket description is how Boot loads its resources - ie the ordering of them. This is unrelated to that. If we wanted to expose our yaml properties as YamlPropertySource (as described [here](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#howto-customize-the-environment-or-application-context)) that would start to matter as we would have to figure out where we want to expose them. As it is currently, this is external to that.

Thoughts on dropping the extra flexibility of the config file locations?

@sbespalov 
@fuss86 
@carlspring

**Comment by carlspring** [comment]
It's a bit confusing, I know.

We have the `strongbox-distribution` and we have the `strongbox-web-core`. The `strongbox-distribution` is the final binary. The `strongbox-web-core` is just the REST API. Typically, we use the `strongbox-web-core` for development and rarely have to debug using the `strongbox-distribution`.

I think it's important to be able to specify the location of the respective yaml files by having a property for them.

The `$HOME/etc/conf/strongbox-security-users.yaml` and `etc/conf/strongbox-security-users.yaml` paths are the defaults based on where you're running things from.

I would suggest keeping these as they are, unless you have a better suggestions for them (and we are indeed open to suggestions).

**Comment by bono007** [comment]
> The $HOME/etc/conf/strongbox-security-users.yaml and etc/conf/strongbox-security-users.yaml paths are the defaults based on where you're running things from.

Ok so that makes sense why the former is tried first and if no file is there is uses the latter. 

Can you tell me where each one is the default for? Its possible that there is another property that can be used to derive the default to use based on the "mode" (aka where its being run from) instead of having a custom loader.

**Comment by fuss86** [comment]
Hi @bono007 :)

Thank you for your comments. As you can see it looks overcomplicated and we would like to simplify it. At final stage we want to get rid of `ConfigurationResourceResolver`. In the time this project was started there was no `Spring Boot` (at least in the `strongbox` project it was not a part of initial technology stack). We decided to migrate to `Spring Boot` couple of months ago. Since then we've been migrating step by step to use `Spring Boot` features. This is why this task was created for.

At the final stage we want have one property value and one default value per each `yaml` configuration file we use. `@Value("${strongbox.users.config.yaml:${home}/etc/conf/strongbox-security-users.yaml}")` is the solution we like as long it works for:
* integration tests
* spring boot maven plugin (spring-boot:run)
* strongbox distribution
All the `yaml` files should be external (not `classpath`) files by default. We change them in runtime and want them to be copyable.

This is the refactoring task. We want to keep the features we have. We want to make it simplier by using `Spring Boot` features. Having things `Spring Boot` way it will be easier for users working with `strongbox` as it will not require additional learning.

I hope it clarifies things a bit more. Thank you for your engagement :)

**Comment by bono007** [comment]
Hi @fuss86 . Yw and ty for you reply.

Yep I got all of that - I totally get the context and evolution. Also, I actually have what you described implemented on my user branch. My main question is do we need those 2 level of defaults that is currently implemented? @carlspring comment makes its sound like it is required. Your's sounds like its not a hard requirement. 

I will continue on w/ this simple approach and run integration tests, `spring-boot:run` and the distribution and see what shakes out of it. 

I will post back results.

**Comment by carlspring** [comment]
@bono007 ,

My comments were more for your better understanding of why this looks like this. You can try it like @fuss86 has explained. If it works, we'll review it and see where it leads us. :)

**Comment by bono007** [comment]
I have a working solution that covers the above cases and looks like this in practice... (this is the entire UsersFileManager.java w/ the changes)....
```java
package org.carlspring.strongbox.users;

import org.carlspring.strongbox.users.dto.UsersDto;
import org.carlspring.strongbox.yaml.YAMLMapperFactory;
import org.carlspring.strongbox.yaml.YamlFileManager;

import javax.inject.Inject;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.FileSystemResource;
import org.springframework.core.io.Resource;
import org.springframework.stereotype.Component;

/**
 * @author Przemyslaw Fusik
 * @author Pablo Tirado
 */
@Component
public class UsersFileManager
        extends YamlFileManager<UsersDto>
{
    @Value("#{@yamlPath.resolve('strongbox.users.config.yaml','etc/conf/strongbox-security-users.yaml')}")
    private Resource resource;

    @Inject
    public UsersFileManager(YAMLMapperFactory yamlMapperFactory)
    {
        super(yamlMapperFactory);
    }

    @Override
    protected Resource getResource()
    {
        return resource;
    }
}
```

Couple of things you will notice. 

1. The Resource is loaded "automagically" using the `@Value` annotation to choose its path to the resource. This relies on the naming of the paths as described [here](https://docs.spring.io/spring/docs/3.2.x/spring-framework-reference/html/resources.html#resources-resourceloader). 

2. The path has some strange SPEL in it - `@yamlPath.resolve`. **WHY?** -- Because of the previous point, the path must be prefixed w/ the proper "type" (`file:`, `classpath:`) for it to choose the proper resource. If we want the SpringBoot auto-magic this has to be done. 

That custom SPEL function simply hides the creation of the final path string. If the custom property is set it will use that and if it does not already begin w/ `file: | classpath:` it will prepend it with `file:`. Otherwise if the custom property is not set it will use the default and prepend it with `file:${strongbox.home}` in the normal mode OR  `classpath:` in test mode.

So while this really cleans up the code that is on the surface (ie the YamlManager and its kids) that complexity of "which file path to use" is still there - just hidden a bit below the surface. I was hoping that Boot did a little more smarts/magic around which type of resource it would load. I am thinking this is still worth doing but will wait until I see the final PR in. I still have a few kinks I am ironing out and a few things I want to try to remove one more piece. 

**Net-net:** making good progress and getting close to a submission. 

Also I will be at SpringOne tomorrow 10/7-10/10 and will make sporadic progress between now and then. However, I will try to get a PR in soon to get some early eyes on it.

**Comment by carlspring** [comment]
Can this now be closed?

**Comment by bono007** [comment]
Let me get that other PR in we are talking about - its related to the cleanup

**Comment by carlspring** [comment]
Okay, cool! Thanks!

**Comment by bono007** [comment]
@carlspring this can be closed now.

**Comment by carlspring** [comment]
Thanks for improving how this works, @bono007 ! :)
